### PR TITLE
Adding qualifying_results()

### LINKED ISSUE #43: Add a qualifying_results() function
A ```qualifying_results()``` function would be a great addition to the project. This function should accept year and race round as parameters. The function should return a DataFrame containing:

- qualifying position
- driver names
- driver numbers
- constructors
- qualifying times for Q1, Q2, and Q3.

The function should also accept an optional parameter ```pos```. This can be utilized to obtain qualifying results with a specific finishing position.

Example:

- ```qualifying_results(2022, 3)``` -> Returns qualifying results for the 3rd race in 2022 (API data [here](https://ergast.com/api/f1/2022/3/qualifying))
- ```lap_times(2022, pos=1)``` -> Returns drivers that got pole position in 2022 (API data [here](https://ergast.com/api/f1/2022/qualifying/1))

### PR DESCRIPTION
**Pull request by achiverram28**
## Description

I have created a python file  called ```qualifying_results.py``` which contains the ```QualifyingResults``` class with the following methods:
- ```get_positions()```
- ```get_names()```
- ```get_driver_numbers()```
- ```get_constructors()```
- ```get_qualifying_times() ```

I have created the ```first_3_qualifying_results_2022.json``` for sample data for testing

I have created ```test_ qualifying_results.py``` file to test the correctness of  ```qualifying_results.py``` file.

## Related Issue(s)

Fixes #43  

## User-facing Changes

Users will now be able to get driver qualifying times in a particular race by typing ```fp.qualifying_results(year, race_round)```.

### CODE REVIEW

**Comment by achiverram28** [comment]
Can u please check it

**Review by alec-kr — COMMENTED** [review]
Great work!

**Review by alec-kr — CHANGES_REQUESTED** [review]
Pylint is not happy :(
image

You should remove all the trailing whitespaces as outlined, and also add a class docstring to the new class you created.

File: tests/test_qualifying_results.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        driver_numbers=self.r_obj.get_driver_numbers()
+        assert driver_numbers == ['16', '1', '11']
+    def test_get_constructors(self):
+        """Test the constructor names returned from get_constructors"""
+        constructor_names=self.r_obj.get_constructors()
+        assert constructor_names == ['Ferrari', 'Red Bull', 'Red Bull']
+    def test_get_qualifying_times(self):
+        """Test the 3 qualifying positions returned from get_qualifying_times"""
+        three_times=self.r_obj.get_qualifying_times()
+        assert three_times == [('1:18.881', '1:18.606', '1:17.868'), ('1:18.580', '1:18.611', '1:18.154'), ('1:18.834', '1:18.340', '1:18.240')]
```
Comment (alec-kr) [inline tests/test_qualifying_results.py:30]: This line is too long. Consider adding each new set of times to a new line.

**Review by alec-kr — CHANGES_REQUESTED** [review]

**Comment by achiverram28** [comment]
please check it now

**Review by alec-kr — CHANGES_REQUESTED** [review]
You have to update the table in the README to include the new function. Be sure to include the function call, a short description, and the returned data type.

Your code is also broken :(
Screenshot from 2022-10-05 17-09-39

Based on the error returned above, it appears as though you are trying to get the Q3 time for drivers that didn't actually make it into Q3.

**Comment by achiverram28** [comment]
F33AEE10-6ED5-471D-84DC-FF2973F35ACE_1_105_c
See its working properly

**Comment by alec-kr** [comment]
You are only testing the first 3 entries. The first 3 drivers have obviously made it into all 3 qualifying sessions, therefore your test will pass.

You have clearly not run the ```fp.qualifying_results()``` to test your code. Please do so, and correct the error

**Comment by achiverram28** [comment]
0ACDDD5E-3B38-4A1B-A05F-3A37D49F2717
See here the details of q2 ,q3 is missing in the api

**Comment by alec-kr** [comment]
## Reproducing your error
To reproduce your code error on your system, you can:
- Install your fork of the repo
- Import the package as ```fp``` or anything you like
- Run ```fp.qualifying_results()```

## The Problem
The test you are showing me in the screenshot is only testing the top 3 drivers. The top 3 drivers obviously made it into all qualifying sessions, as seen in your data. However, if you were to test the 3 slowest drivers, you would notice that your test will fail.

Your current code is checking for the keys ```Q1```, ```Q2```, ```Q3```, regardless if these keys exist or not. 

## The Solution
What you should do, is implement a test to check if the key exists, before attempting to get the qualifying times. Have a look at [this code snippet](https://github.com/alec-kr/F1PyStats/blob/main/f1pystats/driver_info.py#L24-L28) from [driver_info.py](https://github.com/alec-kr/F1PyStats/blob/main/f1pystats/driver_info.py). Here, I am checking the JSON data to determine if the ```permanentNumber``` key is in the current data. If the key exists, I append the value stored at that key. However, if the key does not exist, ```None``` is appended to the list.

You can copy and implement this code into your ```qualifying_results.py``` file, in the ```get_qualifying_times()``` function. This function should contain a for loop, looping over ```self.results```, and checking if the keys for each qualifying session exist. If a particular key exists, only then should you get the value stored at that key, and add it to a list, as seen in [```driver_info.py```](https://github.com/alec-kr/F1PyStats/blob/main/f1pystats/driver_info.py#L24-L28)

**Comment by achiverram28** [comment]
Yeah I am implementing , this is was what even I THOUGHT after I saw the error

**Comment by achiverram28** [comment]
Bro can You check it now

**Comment by alec-kr** [comment]
Your new code seems good. I will test the function later in the day.

**Review by alec-kr — CHANGES_REQUESTED** [review]
I have just tested the function @achiverram28. 

I'm not sure why you are returning a tuple inside the DataFrame. 

Screenshot from 2022-10-06 10-27-30

As seen in the photo above, you are just returning a tuple containing qualifying times for all 3 sessions. This is not quite neat, since it will be difficult for users to access this data from the function.

There should be separate headings for each set of qualifying times. You have a single column name stored as ```Q1,Q2,Q3```, when there should be three columns as shown here:

Screenshot from 2022-10-06 10-28-52

**Comment by alec-kr** [comment]
Maybe you can create separate methods in ```QualifyingResults``` to get the times from each session.

Example: ```get_q1_times()```, ```get_q2_times()```, ```get_q3_times()```

In your ```f1pystats.py```, you will create 3 separate variables for qualifying times, which will store the data for the 3 columns containing qualifying times.

**Comment by alec-kr** [comment]
Note: Please ensure you install and test your package with the changes before submitting an updated PR

**Comment by achiverram28** [comment]
Screenshot 2022-10-07 at 4 47 28 PM
Hope this is fine
And I have tested , I did not get any further errors
I am making a PR now

**Comment by alec-kr** [comment]
I will test it locally later.

**Review by alec-kr — APPROVED** [review]
Tested and works well.
Feature functions as expected, so I will be merging it shortly.

Thank you for contributing 😁

**Pull request merged**